sysfs-led-vince.c

```diff
@@ -0,0 +1,372 @@
+static void        led_control_vince_value_cb      (void *data, int r, int g, int b);
+static void        led_control_vince_close_cb      (void *data);
+
+bool               led_control_vince_probe         (led_control_t *self);
+
+/* ========================================================================= *
+ * ONE_CHANNEL
+ * ========================================================================= */
+
+static void
+led_channel_vince_init(led_channel_vince_t *self)
+{
+    self->cached_max_brightness = sysfsval_create();
+    self->cached_brightness     = sysfsval_create();
+    self->cached_blink          = sysfsval_create();
+
+    self->control_blink = false;
+    self->control_value = 0;
+}
+
+static void
+led_channel_vince_close(led_channel_vince_t *self)
+{
+    sysfsval_delete(self->cached_max_brightness),
+        self->cached_max_brightness = 0;
+
+    sysfsval_delete(self->cached_brightness),
+        self->cached_brightness = 0;
+
+    sysfsval_delete(self->cached_blink),
+        self->cached_blink = 0;
+}
+
+static bool
+led_channel_vince_probe(led_channel_vince_t *self,
+                        const led_paths_vince_t *path)
+{
+    bool ack = false;
+
+    /* Probe control files in reverse existance likelihood order.
+     * Practically all led control directories have 'brightness'
+     * file, most have 'max_brightness' while only some have 'blink'.
+     */
+    if( !sysfsval_open_rw(self->cached_blink, path->blink) )
+        goto cleanup;
+
+    if( !sysfsval_open_ro(self->cached_max_brightness, path->max_brightness) )
+        goto cleanup;
+
+    /* The 'max_brightness' seems to be dynamic. Make an attempt
+     * to set it to an artificially high value and assume that kernel
+     * side will cap the value to the true maximum... */
+#if 1 // TODO: make a QUIRK out of this
+    sysfsval_set(self->cached_max_brightness, 255);
+#endif
+    sysfsval_refresh(self->cached_max_brightness);
+
+    if( sysfsval_get(self->cached_max_brightness) <= 0 )
+        goto cleanup;
+
+    if( !sysfsval_open_rw(self->cached_brightness, path->brightness) )
+        goto cleanup;
+
+    ack = true;
+
+cleanup:
+
+    /* Always close the max_brightness file */
+    sysfsval_close(self->cached_max_brightness);
+
+    /* On failure close the other files too */
+    if( !ack ) {
+        sysfsval_close(self->cached_brightness);
+        sysfsval_close(self->cached_blink);
+    }
+
+    return ack;
+}
+
+static void
+led_channel_vince_set_value(led_channel_vince_t *self,
+                            int value)
+{
+    /*
+     * Since Xiaomi implemented 4095 level brightness in panel dtsi
+     * while caf only have 255 level, we multiply 16 to brightness by default.
+    */
+
+    value *= 16;
+    value = (value < 0) ? 0 : (value < 255) ? value : 255;
+    // value = led_util_scale_value(value,
+    //                             sysfsval_get(self->cached_max_brightness));
```

Comment:
AFAICT this is basically copy of f5121 backed with the only significant change being the max brightness handling right here - is that right?

Reply:
Yes, this is a copy of f5121 years ago, i forget to pr it. Now i have a new vince, and continue to porting.